File: Novac.c
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <stdbool.h>
#include <string.h>

bool Debug = false;
bool Compiled = false;

#include "include/Base.h"
#include "include/lexer.h"
#include "include/Parser.h" 

char *Outname = "Default";

int Exit_With_Error(const char *m, FILE *f, char *buffer)
{
    if (f != NULL)
        fclose(f);

    if (buffer != NULL)
        free(buffer);

    fprintf(stderr, "[-] ERROR: %s\n", m);
    return EXIT_FAILURE;
}

int main(int argc, char *argv[])
{
    if (argc < 2) {
        return Exit_With_Error("No Input File\n", NULL, NULL);
    }

    int FileCount = 0;
    File file = {0};

    for (int i = 1; i < argc; i++) {

        if (strcmp(argv[i], "-o") == 0) {
            if (i + 1 >= argc) {
                return Exit_With_Error("No OutName After -o\n", NULL, NULL);
            }

            if (argv[i + 1][0] == '-') {
                return Exit_With_Error("Invalid OutName After -o\n", NULL, NULL);
            }

            Outname = argv[++i];
        }

        else if (strcmp(argv[i], "--compile") == 0) {
            if (i + 1 < argc) {
                return Exit_With_Error(
                    "Flags Like --compile Need To Be At The End\n",
                    NULL,
                    NULL
                );
            }

            Compiled = true;
        }

        else if (strcmp(argv[i], "--debug") == 0) {
            Debug = true;
        }

        else if (strcmp(argv[i], "--help") == 0) {
            printf("TODO: Give & Open Documentation Link (Local)\n");
            return EXIT_SUCCESS;
        }

        else if (strcmp(CheckFilext(argv[i]), ".nv") == 0) {
            if (FileCount >= 1) {
                return Exit_With_Error(
                    "Too many Input Files, Try Importing instead\n",
                    NULL,
                    NULL
                );
            }

            file.file_path = argv[i];
            Open_Read_File(&file);

            FileCount++;
        }

        else {
            return Exit_With_Error(
                "Unknown Command or Invalid File Type, Try --help\n",
                NULL,
                NULL
            );
        }
    }

    if (FileCount == 0) {
        return Exit_With_Error("No Input File\n", NULL, NULL);
    }

    if (Compiled) {
        m_index = 0;
        Line_count = 1;
        col_count = 1;

        Tokenize(&m_index, &file);
        Parse(tokens, Token_count);
    }

    if (Debug) {
        printf("DEBUG: Input File: %s\n", file.file_path);
        printf("DEBUG: File Length: %zu\n", file.length);
        
        if (strcmp(Outname, "Default") == 0) {
            printf("DEBUG: No Output File Name Enter Set To Default\n");
        } else {
            printf("DEBUG: Output File Name: %s\n", Outname);   
        }
        
        printf("DEBUG: Compiled: %s\n", Compiled ? "true" : "false");

        if (Compiled) {
            printf("DEBUG: Total Tokens: %d\n", Token_count);

            for (int j = 0; j < Token_count; j++) {
                printf(
                    "Token %d -> Type: %s, Value: %d, Row: %d, Col: %zu\n",
                    j,
                    token_type_to_string(tokens[j].type),
                    tokens[j].value,
                    tokens[j].row,
                    tokens[j].col
                );
            }
            
            printf("\n--- EXPE TREE ---\n");
            Print_EXPE(temp->E, 0);
        }
    }

    return EXIT_SUCCESS;
}

File: src/Novac.ts
import { Token_count, Tokenize, Token } from "./lexer.ts";

const args = Deno.args;
const argc = args.length;

let src: string = "";
let file_path: string = "";
let file_length: number = 0;
let Outname: string = "Program";
let tokens: Array<Token> = [];

let Debug = false;
let Compiled = false;

function Exit_With_Error(Message: string): never {
  console.error("[-] ERROR:", Message);
  Deno.exit(1);
}

function print(Message: string) {
  console.log(Message);
}

if (argc === 0) {
  Exit_With_Error("No Input File");
}

// --help
if (args.includes("--help")) {
  print("TODO: Open Local Docs");
}

// --debug
if (args.includes("--debug")) {
  Debug = true;
}

// --compile
if (args.includes("--compile")) {
  const compileIndex = args.indexOf("--compile");

  if (compileIndex !== args.length - 1) {
    Exit_With_Error("--compile needs to be at the end");
  }

  Compiled = true;
}

// -o <NAME>
const outputIndex = args.indexOf("-o");

if (outputIndex !== -1) {
  if (outputIndex + 1 >= args.length) {
    Exit_With_Error("-o requires an output name");
  }

  Outname = args[outputIndex + 1];
}

// .nv file
const file = args.find(arg => arg.toLowerCase().endsWith(".nv"));

if (file === undefined) {
  Exit_With_Error("No .nv File Specified");
}

// Read Nova source
src = await Deno.readTextFile(file);

// Get file information
const info = await Deno.stat(file);

file_path = file;
file_length = info.size;

// Debug information
if (Debug === true) {
  print(`INFO: File Path: ${file_path}`);
  print(`INFO: File Length: ${file_length}`);
  if (Outname === "Program") {
    print("INFO: No FileName, Set To Program");
  }
  print(`INFO: Output Name: ${Outname}`);
  print(`INFO: Compiled: ${Compiled}`);
  print("INFO: Debug: true");
}

// Tokenize
tokens = Tokenize(src, 0);

// Debug token output
if (Debug === true) {
  print(`INFO: Token Count: ${Token_count}`);
  print("INFO: Tokens:");
  
  for (const token of tokens) {
    print(`${token.type} | "${token.value}" | Row: ${token.row} | Col: ${token.col}`
    );
  }
}

if (Compiled === true) {
  print(`TODO: Package Everything Together -> ${Outname}`);
}
